### ISSUE TITLE
Converting the mesh to convex shape

### ISSUE THREAD

**Issue by jkw0701** [flexible-collision-library/fcl#454]
Hello. 

I calculated the two closest points and the distance between the two meshes(obj file)
by using 'MeshDistanceTraversalNode'.

However, I'm wondering if it is able to speed up the calculation when the meshes are converted into the primitive shapes like fcl::Convex.

If possible, how can I convert? I can't any example from the source.
Or, if there is any fast method to calculate the closests points and the distance between mehses?

Thanks in advance.

**Comment by SeanCurtis-TRI** [comment]
1. There isn't any example of the conversion.
2. The `Convex` geometry type is currently limited. It doesn't do any validation on the data provided it so it is distinctly possible to provide it data that is *non*-convex and it will silently proceed.
3. If your mesh is *properly* convex, you may still end up with some redundancy if you have planar triangles. The mesh operates on triangles. The convex type can operate on planar polygons. If you have an underlying mesh that has, for example, planar quads that have been split to be compatible with the mesh type, then simply adding a plane for each triangle will produce an host of redundant plane specifications.

So, that's the bad news. However, if you're confident your mesh is convex and you've merged contiguous, planar triangles, then the documentation in the [`Convex` geometry type](https://github.com/flexible-collision-library/fcl/blob/master/include/fcl/geometry/shape/convex.h#L49-L79) should be enough to help you produce the new answer.

In essence they are:
  1. Collect up your unique mesh vertices.
  2. Count up the number of polygonal faces you have (i.e., if two adjacent triangles are co-planar, turn them into a single quad).
  3. [Encode each face into a massive set of faces](https://github.com/flexible-collision-library/fcl/blob/master/include/fcl/geometry/shape/convex.h#L125-L129)

The most challenging part of that will be the encoding of the faces. It's basically a sequence of integers such as:

```
3 1 10 2
4 3 5 8 9
...
```

Each row above indicates a single polygon. The first `int` indicates the number of vertices. The next `n` ints are indices into your vector list. You must order them in a counter-clockwise order (looking at the polygon from the "outside"). Although, in the illustration, I've broken the sequence of `int`s apart to make a point, in memory it would look like:

`3 1 10 2 4 3 5 8 9 ...`

**Comment by jmirabel** [comment]
@SeanCurtis-TRI gave a complete answer regarding the current status of FCL.
In our [fork](https://github.com/humanoid-path-planner/hpp-fcl/), I worked a bit on class Convex to ensure it works correctly (with the internal GJK / EPA implementation, not with libccd). Here is what I did to make it work.
1. The actual class Convex do not need to know what are the polygons. It only uses the vertices.
2. If you plan to use class Convex for models with many points, then the performances will be dramatic. The complexity of the support mapping implemented in FCL is linear wrt the number of points. It is easy to modify to implement an algorithm of complexity sqrt(N). In this requires, class Convex must store the vertices (and not the polygons). It is also possible (though a lot harder) to write a log(N) support mapping.
- As @SeanCurtis-TRI, if you know that your model is convex, it is very to create a Convex object.

All the above points may not be true when using libccd.
